beacon-chain/endtoend/endtoend_test.go

```diff
@@ -0,0 +1,514 @@
+package endtoend
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/big"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/pborman/uuid"
+	ev "github.com/prysmaticlabs/prysm/beacon-chain/endtoend/evaluators"
+	contracts "github.com/prysmaticlabs/prysm/contracts/deposit-contract"
+	"github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/testutil"
+	"google.golang.org/grpc"
+)
+
+type end2EndConfig struct {
+	minimalConfig  bool
+	tmpPath        string
+	epochsToRun    uint64
+	numValidators  uint64
+	numBeaconNodes uint64
+	contractAddr   common.Address
+	evaluators     []ev.Evaluator
+}
+
+type beaconNodeInfo struct {
+	processID   int
+	datadir     string
+	rpcPort     uint64
+	monitorPort uint64
+	grpcPort    uint64
+	multiAddr   string
+}
+
+type validatorClientInfo struct {
+	processID   int
+	monitorPort uint64
+}
+
+func runEndToEndTest(t *testing.T, config *end2EndConfig) {
+	tmpPath := path.Join("/tmp/e2e/", uuid.NewRandom().String()[:18])
+	if err := os.MkdirAll(tmpPath, os.ModePerm); err != nil {
+		t.Fatal(err)
+	}
+	config.tmpPath = tmpPath
+	fmt.Printf("Test Path: %s\n", tmpPath)
+
+	contractAddr, keystorePath, eth1PID := startEth1(t, tmpPath)
+	config.contractAddr = contractAddr
+	beaconNodes := startBeaconNodes(t, config)
+	valClients := initializeValidators(t, config, keystorePath, beaconNodes)
+
+	beaconLogFile, err := os.Open(path.Join(tmpPath, "beacon-0.log"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := WaitForTextInFile(beaconLogFile, "Chain started within the last epoch"); err != nil {
+		t.Fatal(err)
+	}
+	t.Log("Chain has started")
+
+	if config.numBeaconNodes > 1 {
+		t.Run("peers_connect", func(t *testing.T) {
+			for _, bNode := range beaconNodes {
+				if err := PeersConnect(bNode.monitorPort, config.numBeaconNodes-1); err != nil {
+					t.Fatalf("failed to connect to peers: %v", err)
+				}
+			}
+		})
+	}
+
+	conn, err := grpc.Dial("127.0.0.1:4000", grpc.WithInsecure())
+	if err != nil {
+		t.Fatalf("fail to dial: %v", err)
+	}
+	time.Sleep(2 * time.Second)
```

Comment:
Sleeping can be flaky, can you write something like `waitUntilGRPCServerIsReady(t)` where it will wait (up to some timeout) for the test to proceed.